### ISSUE TITLE
kube-controller-manager crashloop backup when assigning an existing public ip to a loadbalancer service

### ISSUE THREAD

**Issue by PascalVA** [Azure/aks-engine#4417]
First of all I am not sure if this is the right place to post this issue. I have the issue on aks-engine, but it could be the cloud controller manager.

When deploying a new aks-engine cluster (with more than 1 master node), if we pre-create a public ip address in the same resource group as aks-engine, and then create a loadbalancer service in Kubernetes to use it, kube-controller-manager goes into a Nil pointer deref crashloop backoff (until you delete the ip address from the resource group).

If you use a dynamically created ip, it seems to work, unless you delete the loadbalancer service and want to re-use that dynamic ip address again later.

**AKS Engine Version***
0.62.0

**Kubernetes Version**
1.21.0, 1.20.5

**To Reproduce**

1. log in to az cli and set the correct test subscription
1. setup temporary dir and download aks-engine release 0.62.0
    ```bash
    mkdir aks-engine-test && cd aks-engine-test
    curl -L https://github.com/Azure/aks-engine/releases/download/v0.62.0/aks-engine-v0.62.0-linux-amd64.tar.gz | tar xzf -
    ```
1. precreate resource group and public ip (save the ip to apply it later in the LB)
    ```bash
    az group create --name k8s-1-21-0-aksengine --location eastus
    az network public-ip create --name testlbip --resource-group k8s-1-21-0-aksengine --allocation-method static --sku Standard
    ```
1. create the aks-engine cluster (it is important you keep 3 master replicas)
    * `kubernetes.json` content (replace the service principal)
        ```json
        {
            "apiVersion": "vlabs",
            "properties": {
                "orchestratorProfile": {
                    "orchestratorType": "Kubernetes",
                    "orchestratorRelease": "1.21",
                    "orchestratorVersion": "1.21.0",
                    "kubernetesConfig": {
                        "useManagedIdentity": false
                    }
                },
                "masterProfile": {
                    "count": 3,
                    "distro": "aks-ubuntu-18.04",
                    "dnsPrefix": "k8s-1-21-0-ld1meeorpm1",
                    "vmSize": "Standard_D4s_v3"
                },
                "agentPoolProfiles": [
                    {
                        "count": 1,
                        "distro": "aks-ubuntu-18.04",
                        "name": "agentpool",
                        "vmSize": "Standard_D4s_v3"
                    }
                ],
                "linuxProfile": {
                    "adminUsername": "testuser",
                    "ssh": {
                        "publicKeys": [
                            {
                                "keyData": "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC+ou4eENCIYGWNg9jr8eYvov7yVg9dMiceZTsmehVWLHu+gbbik4tsE2PNykun/aHX7rVG76w0+dA/m3lZJrwYaU3ZurU8apwquv2WJLfxPlNlKLnScgOzk9NV++sXLHgKX2L1uM/AyBDomNBannwfqqzfpmcRtjM0euwR04fpRZuqlsZCLwpblGTaKmdk9EKN4/M5iZn8EaJfUokz2XcU+WEtvUgZCoUKIZvT8TLA387rTUk5ISQcQMRHhczmPT0Eg/gMlPBF/dPIiyR5Np2G0lgfR+sNbjLLI9TH2qR79FiqQfn6lFnz+agTWskOWSn16mASYXYmjOHnJehUVvN0Arbd21gP5fsDybEepDsDFUofssdYQeo9744xzvLrT8eqUbMx/LAvue5p+deRR8hTP27r88NQ9OzZBwv2MCbUScPr/f3XKI4nfSlFN+J22mlbQ7E3m2N/pERQT0CT+IIZeJ+W3IoMttupKWGXyLmEnbZng9pmiH+9nQkqAi+xGuE="
                            }
                        ]
                    }
                },
                "servicePrincipalProfile": {
                    "clientId": "CLIENT_ID",
                    "secret": "CLIENT_SECRET"
                }
            }
        }
        ```
    * generate the aks-engine deployment templates
        ```bash
        ./aks-engine-v0.62.0-linux-amd64/aks-engine generate --output-directory aksengine-output --api-model kubernetes.json
        ```
    * deploy the aks-engine templates with azure group deployment (not aks-engine)
        ```bash
        az deployment group create --name aksengine --resource-group k8s-1-21-0-aksengine --template-file ./aksengine-output/azuredeploy.json --parameters ./aksengine-output/azuredeploy.parameters.json
        ```
1. create the loadbalancer service (change IP_ADDRESS to the ip of your resource)
    ```bash
    cat <<EOF | kubectl --kubeconfig aksengine-output/kubeconfig/kubeconfig.eastus.json apply -f -
    apiVersion: v1
    kind: Service
    metadata:
      name: test-app
    spec:
      selector:
        app: test-app
      ports:
        - protocol: TCP
          port: 80
          targetPort: 80
      loadBalancerIP: IP_ADDRESS
      type: LoadBalancer
    EOF
    ```
1. check the logs of the controller manager
    ```bash
    kubectl --kubeconfig aksengine-output/kubeconfig/kubeconfig.eastus.json -n kube-system logs -l component=kube-controller-manager -f
    ```

You will see the following output:
```bash
I0505 10:44:43.434034       1 azure_backoff.go:282] LoadBalancerClient.List(k8s-1-21-0-aksengine) success
I0505 10:44:43.443629       1 azure_backoff.go:282] LoadBalancerClient.List(k8s-1-21-0-aksengine) success
I0505 10:44:43.470387       1 azure_backoff.go:298] PublicIPAddressesClient.List(k8s-1-21-0-aksengine) success
I0505 10:44:43.476785       1 azure_backoff.go:298] PublicIPAddressesClient.List(k8s-1-21-0-aksengine) success
I0505 10:44:43.503663       1 azure_backoff.go:298] PublicIPAddressesClient.List(k8s-1-21-0-aksengine) success
E0505 10:44:43.503818       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 2467 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x40c2640, 0x72b6590)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x95
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x86
panic(0x40c2640, 0x72b6590)
    /usr/local/go/src/runtime/panic.go:965 +0x1b9
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).serviceOwnsFrontendIP(0xc000198a00, 0x0, 0xc00027eee0, 0xc00134b6c0, 0xc00134b6e0, 0xc00134b6f0, 0x0, 0xc00134b6d0, 0xc00107d500, 0x60920000, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_standard.go:370 +0x2a2
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).getServiceLoadBalancerStatus(0xc000198a00, 0xc00107d500, 0xc001ed2910, 0x0, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:511 +0x3a5
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).getServiceLoadBalancer(0xc000198a00, 0xc00107d500, 0x7ffd2fd0d6d0, 0x16, 0xc000b7c3a0, 0x3, 0x4, 0x1, 0xc000b7c3e0, 0x2, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:383 +0x3e5
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).reconcileLoadBalancer(0xc000198a00, 0x7ffd2fd0d6d0, 0x16, 0xc00107d500, 0xc000b7c3a0, 0x3, 0x4, 0x1, 0x0, 0x0, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:1091 +0x409
k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure.(*Cloud).EnsureLoadBalancer(0xc000198a00, 0x5068448, 0xc0001a2008, 0x7ffd2fd0d6d0, 0x16, 0xc00107d500, 0xc000b7c3a0, 0x3, 0x4, 0x0, ...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/legacy-cloud-providers/azure/azure_loadbalancer.go:187 +0x365
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).ensureLoadBalancer(0xc0004d0000, 0xc00107d500, 0x0, 0x0, 0x6)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:452 +0xf3
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).syncLoadBalancerIfNeeded(0xc0004d0000, 0xc00107d500, 0xc001a937c0, 0x10, 0xc000b7c360, 0xc0024a2cd0, 0x15a5a93)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:407 +0x9ab
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).processServiceCreateOrUpdate(0xc0004d0000, 0xc00107d500, 0xc001a937c0, 0x10, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:343 +0xdf
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).syncService(0xc0004d0000, 0xc001a937c0, 0x10, 0x0, 0x0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:854 +0x325
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).processNextWorkItem(0xc0004d0000, 0x203000)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:301 +0xf5
k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).worker(...)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:280
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc00249dc80)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00249dc80, 0x5007fe0, 0xc001b16420, 0x4b22501, 0xc0007038c0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x9b
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00249dc80, 0x3b9aca00, 0x0, 0x1, 0xc0007038c0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc00249dc80, 0x3b9aca00, 0xc0007038c0)
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
created by k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service.(*Controller).Run
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/cloud-provider/controllers/service/controller.go:238 +0x24e
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x32b9022]
```

**Expected behavior**
The loadbalancer is ensured, and our service is reachable.

**Please don't remove the text below**
Notify @az-policy-kube

**Comment by jackfrancis** [comment]
@feiskyer @andyzhangx Can you confirm that this is a configuration that Azure cloud provider can't handle? Or is this a bug that we want to address in out-of-tree cloud provider code? Thanks!

**Comment by feiskyer** [comment]
Thanks for reporting the issue. This is caused by a bug and would be fixed by https://github.com/kubernetes/kubernetes/pull/101739. The cherry-pick PRs have been sent for 1.20/1.21 branches, and they're pending on approval now.

**Comment by PascalVA** [comment]
Thanks @feiskyer, I did come across this issue but was not sure it was fully addressing the issue, thanks for clarifying that, I'll be happy to see a fix come in.

**Comment by jackfrancis** [comment]
Thanks @feiskyer Can @PascalVA try a patched azure cloud provider now using a build of cloud-controller-manager?

**Comment by PascalVA** [comment]
@jackfrancis thanks, there is no need, we decided to postpone our upgrade a bit until it is working, I can wait for the release.

**Comment by Michael-Sinz** [comment]
We really need this at it prevents us from running 1.20 or 1.21 clusters!

**Comment by feiskyer** [comment]
the fix is being cherry-picked to 1.20 and 1.21 by https://github.com/kubernetes/kubernetes/pull/101886 and https://github.com/kubernetes/kubernetes/pull/101885, and they would be part of 1.20.8 and 1.21.2.

**Comment by PascalVA** [comment]
We have just deployed a Kubernetes cluster of version 1.21.2 and the problem of the nil pointer dereference seem to be sloved indeed. However, the functionality we wanted (pre-creating an ip and using it on the loadbalancer) is still not working. I will create a new issue for this.
